### PR TITLE
Optimize pg_get_expr to avoid unnecessary lock on relation. 

### PR DESCRIPTION
**Pull request by JunfengYang**
This interface is designed to parse a generic rule expression and rule
expressions contain references to relations. Acquiring AccessShareLock
on a relation encountered while deparsing a rule expression seems
logical.
For partition tables, the RANGE and LIST expressions are much simpler -
they cannot contain any references to relations.
So we decide to optimize out the lock when the expression contains
no Var nodes.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/utils/adt/ruleutils.c [hunk 1]
```diff
@@ -1893,6 +1893,48 @@ pg_get_expr_ext(PG_FUNCTION_ARGS)
...
+{
+	Assert(var_exist);
+
+	if (node == NULL)
+		return false;
+
+	if (nodeTag(node) == T_Var)
+	{
+		*var_exist = true;
+		return false;
```
Comment (zhangh43) [inline src/backend/utils/adt/ruleutils.c:1910]: should return true here to quit the walk?

Comment (JunfengYang) [reply src/backend/utils/adt/ruleutils.c:1910]: Ah you are right!

File: src/backend/utils/adt/ruleutils.c [hunk 2]
```diff
@@ -1893,6 +1893,48 @@ pg_get_expr_ext(PG_FUNCTION_ARGS)
...
+	{
+		*var_exist = true;
+		return true;
+	}
+
+	return expression_tree_walker(node, find_var, var_exist);
+}
+
+/*
+ * If there are no Var nodes exists in the expression tree, we treate
```
Comment (asimrp) [inline src/backend/utils/adt/ruleutils.c:1917]: typo: treat

Comment (asimrp) [reply src/backend/utils/adt/ruleutils.c:1917]: Let's augment this comment to include the fact that avoiding a lock on a relation is essential when this function is invoked by `pg_partitions` view.  The view takes locks on every partition table defined in the database and that conflicts heavily with concurrent DDL transactions, if any.

File: src/backend/utils/adt/ruleutils.c [hunk 3]
```diff
@@ -1893,6 +1893,48 @@ pg_get_expr_ext(PG_FUNCTION_ARGS)
...
+ */
+static bool
+find_var(Node *node, bool *var_exist)
+{
+	Assert(var_exist);
+
+	if (node == NULL)
+		return false;
+
+	if (nodeTag(node) == T_Var)
```
Comment (asimrp) [inline src/backend/utils/adt/ruleutils.c:1907]: We need to make this test stronger, so that it matches only expressions that are similar to a partition list or range spec.  We know the permissible node types in partition specs.  It seems better to change this function to assert absence of all the nodes other than those permitted in partition specs.  E.g. if `Const` is the only node permitted in partition spec, need to change this condition to:
```C
    if (nodeTag(node) != T_Const)
    {
        *simple = false;
        return false;
    }
```

Comment (asimrp) [reply src/backend/utils/adt/ruleutils.c:1907]: Or, may be, rename the `is_simple_expr()` to `is_partition_spec_expr()` to clarify the intention.

Comment (JunfengYang) [reply src/backend/utils/adt/ruleutils.c:1907]: Actually, there could be other node types, for example, function. It's safer to only check the Var node here.
Other types will be covered by `expression_tree_walker`.
This optimization could also apply to other cases too.

**Comment by asimrp** [comment]
This simple isolation2 spec should be enough to test this patch:
```SQL
-- Test that pg_partitions view does not lock a table                                                                                                                                                                                         

create table pg_partitions_ddl_tab(a int, b int)
distributed by (a) partition by range (b)
( start (1) end (20) every(5) );

1: begin;
1: lock pg_partitions_ddl_tab in access exclusive mode;
1: lock pg_partitions_ddl_tab_1_prt_2 in access exclusive mode;


-- The following query should not be blocked by session 1 because no                                                                                                                                                                          
-- locks should be held by pg_partitions view.                                                                                                                                                                                                
select tablename, partitiontablename, partitionboundary from pg_partitions where tablename like 'pg_partitions_ddl%';

1: end;
1q:
```

File: src/backend/utils/adt/ruleutils.c (hunk 3 above)
Comment (asimrp) [reply src/backend/utils/adt/ruleutils.c:1907]: Thank you @JunfengYang for explaining this offline.  It's possible to have nodes such as FuncExpr in partition specs.  Which means, the list of all nodes permitted in partition spec is large.  Therefore, it's more convenient to test only the presence of Var nodes.

Regarding the concern that whether the patch introduces regressions by not locking a relation when it should be locked, there is one-one correspondence between Var nodes in an expression and locks acquired in order to deparse the expression.  Moreover, the OID passed as argument to `pg_get_expr()` is locked only for a short duration.  Not acquiring locks on it shouldn't cause any problems.

**Review by asimrp — APPROVED** [review]

File: src/backend/utils/adt/ruleutils.c [hunk 4]
```diff
@@ -1914,27 +1890,53 @@ pg_get_expr_ext(PG_FUNCTION_ARGS)
...
+ * it as a simple expression since deparse_expression_pretty doesn't
+ * need to interpret Vars in the node tree. So skip calling
+ * deparse_context_for, this avoids unnecessary lock on relation.
+ * Currently this optimization only apply to pg_partitions view to
+ * avoid acquire lock on the relation. Otherwise the view takes locks
+ * on every partition table defined in the database and that conflicts
+ * heavily with concurrent DDL transactions, if any.
+ */
+static bool
+is_simple_expr(Node *expr)
```
Comment (x4m) [inline src/backend/utils/adt/ruleutils.c:1927]: I'd suggest calling this function like expr_has_vars() or something like that. Or is the concept of "simple expression" used elsewhere?

File: src/backend/utils/adt/ruleutils.c [hunk 5]
```diff
@@ -1914,27 +1890,53 @@ pg_get_expr_ext(PG_FUNCTION_ARGS)
...
+{
+	bool	containVar = false;
+
+	if (expr == NULL)
+		return true;
+
+	if (nodeTag(expr) == T_Var)
```
Comment (x4m) [inline src/backend/utils/adt/ruleutils.c:1934]: Isn't this case handled within recursion?

**Review by x4m — CHANGES_REQUESTED** [review]

File: src/backend/utils/adt/ruleutils.c (hunk 5 above)
Comment (JunfengYang) [reply src/backend/utils/adt/ruleutils.c:1934]: This just returns directly. `expression_tree_walker` can't walk in it.

**Review by x4m — APPROVED** [review]

**Comment by ashwinstar** [comment]
@d please help review this PR

**Comment by d** [comment]
## TL;DR: LGTM. Ship it.

## Long review
Wow this is **very** clever: it looked wrong at first glance, and it took me a long time (like 30 minutes and 4 strong :coffee: espresso shots) to understand.

I'm a little conflicted with this change and all changes that preceded this:

1. #7646 (and #7478) purportedly solved the "problem" of concurrently dropped tables in regression test disrupting `pg_partitions` by taking a lock on the relation specified by the second argument of `pg_get_expr`.
1. But in hindsight, #7478 and #7646 didn't seem necessary at all given that at the time of proposing them we've already merged postgres/postgres@2ffa740be9d96a37 which takes care of taking a brief lock to guard retrieving column names from the table
1. #7646 (and #7478) also caused other unintended consequences, like breaking the `pg_get_expr` contract, as noted by @leskin-in, @maksm90, and @x4m over on issue #11144 
1. At this point my conclusion would be to just revert #7646 and move on.
1. But this PR addresses a gnarly issue: `pg_partitions` didn't need to use the second argument of `pg_get_expr` at all! But it did anyway -- this causes an extra lock being taken for a likely brief window, except when another session preempts us. It's too late in the product lifecycle to make a change to `pg_partitions`, so without this patch we'd have to let `pg_partitions` wait.

**Comment by asimrp** [comment]
Thank you @d for the long review.  Admittedly, we should have explicitly stated point 2 either in comments or in commit message, so as to save you from those strong espresso shots.
